### PR TITLE
Notable Non-Record: JEPA — 1.4447 BPB — Joint Embedding Predictive Architecture for LLMs

### PR DESCRIPTION
**Pull request by gowtham0992**
# First JEPA Submission: 1.4447 BPB — Joint Embedding Predictive Architecture for LLMs

LLM-JEPA Auxiliary Loss + Random Token Masking Views + Cosine Similarity + Asymmetric Stop-Gradient + 11L Transformer + int6 GPTQ + sliding window eval

**val_bpb: 1.4447 (seed=42)** | 10.05 MB artifact | 8×H100 SXM, 555s training + 74s eval

## Results (seed=42, 8×H100 SXM)

| Metric | Value |
|--------|-------|
| Sliding BPB | 1.4447 |
| val_bpb | 1.4447 |
| RT bpb | 1.4447 |
| Steps | 1,790 |
| ms/step | 310.09 |
| Training time | 555s |
| Artifact | 10,054,691 bytes (10.05 MB) |
| Parameters | 26,993,788 (same as baseline) |

## Method

Training loss: **L = L_NTP + λ × L_JEPA**

Following LLM-JEPA (Balestriero et al., arXiv:2509.14252), we add a JEPA auxiliary loss that trains the model to predict invariant embeddings across different corruptions of the same sequence.

### View Creation
- **View 1**: input sequence with 15% of tokens replaced by random tokens (mask A)
- **View 2**: same sequence with 15% of tokens replaced by random tokens (mask B)
- Different random masks create different "views" of the same underlying content

### JEPA Loss
1. Each view processed through the full model independently
2. Mean-pooled hidden states serve as sequence embeddings
3. `L_JEPA = 1 - cosine_similarity(emb_view1, emb_view2.detach())`
4. `detach()` on target creates asymmetric objective (prevents collapse without negative pairs)

### Architecture
- Standard 11L causal Transformer (unchanged — JEPA is a training objective, not architecture)
- λ = 1.0 (JEPA loss weight)
- 3 forward passes per step: 1 for NTP + 2 for JEPA views

## Key Techniques

| Technique | Impact | Notes |
|-----------|--------|-------|
| Random token masking views | Creates meaningful view pairs | 15% corruption rate per view |
| Mean pooling | Robust embedding | More stable than last-token for corrupted inputs |
| Stop-gradient on target | Prevents collapse | Asymmetric design from BYOL/I-JEPA |
| Cosine similarity | Scale-invariant | Matches LLM-JEPA paper's recommended metric |

## Why JEPA for Language Models

Standard NTP trains token-level predictions. JEPA adds a complementary embedding-space signal: the model must learn representations where semantically equivalent sequences (same content, different corruptions) map to similar embeddings. The LLM-JEPA paper shows 5-15% downstream improvement across Llama, Gemma, OpenELM, and OLMo families.

## Command

```bash
JEPA_ENABLED=1 \
JEPA_LAMBDA=1.0 \
SEED=42 \
python3 -m torch.distributed.run --nproc_per_node=8 train_gpt.py
```

## Compliance

- [x] Artifact ≤16,000,000 bytes (10,054,691)
- [x] Training ≤600s on 8×H100 SXM (555s)
- [x] Eval ≤600s (74s)
- [x] No validation data during training
- [x] No network calls during evaluation
- [x] No external compute
- [x] Reproducible from `train_gpt.py`

## References

- LLM-JEPA: [arXiv:2509.14252](https://arxiv.org/abs/2509.14252) (Balestriero et al., 2025)
- I-JEPA: [arXiv:2301.08243](https://arxiv.org/abs/2301.08243) (Assran et al., 2023)

## Included Files

- `train_gpt.py` — full training script
- `train_seed42.txt` — training log
- `submission.json` — metadata

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — Notable Non-Record: JEPA — 1.4447 BPB — Joint Embedding Predictive Architecture for LLMs

**BPB:** 1.4447 | **Compliance:** LOOKS CLEAN — pure-neural submission, no TTT/SLOT/n-gram-cache

**What I found in the code** (head SHA `b0ab6670af35`, file `records/track_10min_16mb/2026-03-29_JEPA_LLM/train_gpt.py`):

Static code review found no TTT adaptation function, no SLOT optimization loop, no n-gram-cache class, and no pre-quant val-token fine-tune. The eval path uses the standard sliding-window stride-64 pattern. The submission is a pure-neural architecture iteration on the standard SP1024/SP4096/SP8192 baseline.

**CPU smoke test (CT2038 proteus-engine, 2026-04-11):** import OK in 0.05s, dim=512, layers=11, vocab=1024, code=72358 B, SMOKE_TEST_PASS

**Verdict:** LOOKS CLEAN.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks (3-seed validation, under-16MB artifact cap, ≤600s train + ≤600s eval on 8×H100 SXM). No compliance flags from the classification pass — this looks like a clean pure-neural iteration on the standard baseline.

*Auto-classification caveat:* this review was drafted by the AST-based classifier. If there's a non-standard eval mechanism (logit postprocessing, hedge mixing, etc.) that I missed because it's factored into a helper file or a non-standard function name, please flag it and I'll re-run the audit manually.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). CPU smoke test (CT2038 proteus-engine, 2026-04-11): import OK in 0.05s, dim=512, layers=11, vocab=1024, code=72358 B, SMOKE_TEST_PASS. Classification via deterministic AST-based `classify_prs.py` (pattern bank derived from ~65 manually-reviewed PRs earlier in the 2026-04-11 sweep). This review was auto-drafted from a template and spot-checked before posting — if the template misread your code, please call it out so I can iterate the classifier.*

**Comment by gowtham0992** [comment]
I noticed this was also a notable non-record submission but lived under `records/track_10min_16mb/`.

Moved `records/track_10min_16mb/2026-03-29_JEPA_LLM/` to `records/track_non_record_16mb/2026-03-29_JEPA_LLM/`.

The latest push is rename-only for those six files, with no content changes.
